flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java

```diff
@@ -227,6 +227,20 @@ public void close() throws Exception {
         consumer.close();
     }
 
+    @Override
+    public void pauseOrResumeSplits(
+            Collection<KafkaPartitionSplit> splitsToPause,
+            Collection<KafkaPartitionSplit> splitsToResume) {
+        consumer.resume(
+                splitsToResume.stream()
+                        .map(KafkaPartitionSplit::getTopicPartition)
+                        .collect(Collectors.toList()));
+        consumer.pause(
+                splitsToPause.stream()
+                        .map(KafkaPartitionSplit::getTopicPartition)
+                        .collect(Collectors.toList()));
```

Comment:
I'm not sure how likely is this `resume` to ever block and even if it blocks, I don't think it should cause any problems. Keep in mind that this is not an exact feature. We do not guarantee that no records will be processed . The whole watermark alignment requires sending RPCs between TMs and JM so it has built in delay/lag.

At the same time I don't see how doing this in the other way around would cause much harm. At worst the consumer would have a for a very short period of time no splits to process if everything is paused but not yet resumed.